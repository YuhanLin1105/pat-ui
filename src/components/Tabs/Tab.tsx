import React, { FC } from 'react';
import { classNames } from '../../utils/classNames';

export interface ITabProps {
  value?: any;
  label?: any;
  wrapped?: boolean;
  disabled?: boolean;
  className?: string;
}
export const Tab: FC<ITabProps> = (props) => {
  const { children, className, ...rest } = props;

  return <div>{children}</div>;
};

export default Tab;