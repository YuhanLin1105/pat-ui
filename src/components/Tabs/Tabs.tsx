import React, {
  Children,
  FC,
  ReactElement,
  cloneElement,
  useState,
  useEffect,
} from 'react';
import { classNames } from '../../utils/classNames';

export interface ITabsProps {
  /** children must be React Element */
  children?: ReactElement[];
  /** set customized css class */
  className?: string;
  /** set default tab and tab content */
  defaultTab?: string | number;
  /** set vertial tabs */
  vertical?: boolean;
}
/**
 * A tabs container allows user to set multiple tabs and related tab content.
 *
 * ```js
 * import { Tabs, TabsPanel, Tab, TabsContent } from 'pat-ui'
 * ```
 */
export const Tabs: FC<ITabsProps> = (props) => {
  const { children, defaultTab, className, vertical, ...rest } = props;
  //console.log(children);
  const [tabValue, setTabValue] = useState<any>();

  // useEffect(() => {
  //   if (defaultTab) {
  //     setTabValue(defaultTab);
  //   }
  // }, []);

  let styleClasses = classNames('tabs', {
    vertical: !!vertical,
  });
  if (className) {
    styleClasses += ' ' + className;
  }
  return (
    <div className={styleClasses}>
      {children
        ? Children.map(children, (child: ReactElement) =>
            cloneElement(child, { tabValue, vertical, setTabValue })
          )
        : children}
    </div>
  );
};

export default Tabs;
