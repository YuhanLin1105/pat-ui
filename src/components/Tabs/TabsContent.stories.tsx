// import React from 'react';
// import { action } from '@storybook/addon-actions';
// import TabsPanel from './Tabs';
// import Tab from './Tabs';
// import TabsContent from './TabsContent';

// export default {
//   title: 'TabsContent',
//   component: TabsContent,
// };

// export const tabsContent = () => (
//   <div>
//     <TabsContent>content</TabsContent>
//   </div>
// );
