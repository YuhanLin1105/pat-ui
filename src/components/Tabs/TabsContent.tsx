import React, {
  FC,
} from 'react';
import { classNames } from '../../utils/classNames';

export interface ITabsContentProps {
  index?: string | number;
  tabValue?: string;
  className?: string;
}
export const TabsContent: FC<ITabsContentProps> = (props) => {
  const { children, className, tabValue, index, ...rest } = props;

  let styleClasses = classNames('tabs__content', {
  });
  if (className) {
    styleClasses += ' ' + className;
  }
  if (tabValue === index){
    styleClasses += ' actived';
  }
    return (
      <div
        className={styleClasses}
      >
        {children}
      </div>
    );
};

export default TabsContent;
