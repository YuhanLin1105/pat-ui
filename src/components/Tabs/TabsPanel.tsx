import React, {
  Children,
  createRef,
  FC,
  ReactElement,
  useEffect,
  useState,
} from 'react';
import { classNames } from '../../utils/classNames';
import Button from '../Button/Button';

export type PanelType = 'primary' | 'default';

export interface ITabsPanelProps {
  /** children must be React Element */
  children?: ReactElement[];
  /** set tabsPanel type */
  type?: PanelType;
  /** set centered tabsPanel */
  centered?: boolean;
  /** set scrollable tabsPanel */
  scrollable?: boolean;
  /** set vertical tabsPanel */
  vertical?: boolean;
  /** set customized css class */
  className?: string;
  /** default tab state provided by tabs */
  tabValue?: string | number;
  /** a callback to provide current value */
  onClick?: (val: any) => void;
  /** set tab value state provided by tabs */
  setTabValue?: (val: any) => {};
}

export const TabsPanel: FC<ITabsPanelProps> = (props) => {
  const {
    children,
    className,
    tabValue,
    type,
    centered,
    scrollable,
    vertical,
    onClick,
    setTabValue,
    ...rest
  } = props;

  let styleClasses = classNames('tabs__panel', {
    [`panel-type-${type}`]: true,
    centered: !!centered,
    [`scrollable-row`]: !!scrollable && !vertical,
    [`scrollable-col`]: !!scrollable && !!vertical,
    vertical: !!vertical,
  });
  if (className) {
    styleClasses += ' ' + className;
  }
  const arrLength = children?.length;
  const [elRefs, setElRefs] = useState([] as any);
  useEffect(() => {
    scrollable &&
      setElRefs((elRefs: any) =>
        Array(arrLength)
          .fill(null)
          .map((_, i) => elRefs[i] || createRef())
      );
  }, [arrLength]);
  //console.log(elRefs);

  const handleTabClick = (e: any, i: number) => {
    if (onClick) {
      onClick(e.target.value);
      setTabValue && setTabValue(e.target.value);
    }
    if (scrollable && elRefs[i].current !== null) {
      elRefs[i].current.scrollIntoView({ behavior: 'smooth', block: 'end' });
    }
  };

  return (
    <div className={styleClasses}>
      {children
        ? Children.map(children, (child: ReactElement, i) => {
            const {
              className,
              value,
              label,
              wrapped,
              disabled,
              ...rest
            } = child.props;

            let btnStyleClasses = classNames('tabs__panel__tab-button', {
              [`tab-type-${type}`]: true,
              wrapped: !!wrapped,
              disabled: !!disabled,
            });
            if (className) {
              btnStyleClasses += ' ' + className;
            }
            if (tabValue === value) {
              if (!disabled) {
                btnStyleClasses +=
                  ' ' + `${vertical ? 'actived-vertical' : 'actived'}`;
              } else {
                setTabValue && setTabValue(null);
              }
            }
            return (
              <div className={`tabs__panel__tab`} ref={elRefs[i]}>
                <Button
                  disabled={disabled}
                  className={btnStyleClasses}
                  value={value}
                  onClick={(e: any) => handleTabClick(e, i)}
                >
                  {label}
                </Button>
              </div>
            );
          })
        : children}
    </div>
  );
};

TabsPanel.defaultProps = {
  type: 'default',
};

export default TabsPanel;
