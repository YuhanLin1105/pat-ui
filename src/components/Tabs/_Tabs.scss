.tabs {
  &__panel {
    display: flex;
    box-shadow: 0px 2px 4px -1px rgba(0, 0, 0, 0.2),
      0px 4px 5px 0px rgba(0, 0, 0, 0.14), 0px 1px 10px 0px rgba(0, 0, 0, 0.12);

    &.panel-type-default {
      @include tabs-panel-type($white);
    }
    &.panel-type-primary {
      @include tabs-panel-type($light-blue);
    }
    &.centered{
        justify-content: center;
    }
    &__tab {
      &-button {
        height: 100%;
        max-width: 260px;
        padding: 10px 20px;
        display: flex;
        justify-content: center;
        align-items: center;
        overflow: hidden;
        text-transform: uppercase;
        &.tab-type-default {
          @include tab-btn-style($gray-500, $light-blue);
        }
        &.tab-type-primary {
          @include tab-btn-style($gray-400, $white);
        }
        &.wrapped {
          word-wrap: break-word;
          white-space: normal;
        }
      }
    }
  }
  &__content {
    //.tabs__content
    display: none;
    background-color: #eee;
    padding: 20px;
    &.actived {
      //.tabs__content.actived
      display: block;
    }
  }
}
