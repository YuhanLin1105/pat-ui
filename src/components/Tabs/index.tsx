 export { default } from './Tabs';

//import{FC}from'react';
