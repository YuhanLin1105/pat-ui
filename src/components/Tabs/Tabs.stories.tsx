import React from 'react';
import { action } from '@storybook/addon-actions';
import TabsPanel from './TabsPanel';
import Tab from './Tab';
import TabsContent from './TabsContent';
import Tabs from './Tabs';

export default {
  title: 'Tabs',
  component: Tabs,
};

export const DefaultTabs = () => (
  <Tabs onClick={action('tabs selected')}>
    <TabsPanel>
      <Tab value="one" label="tab one"></Tab>
      <Tab value="two" label="tab two"></Tab>
    </TabsPanel>
    <TabsContent index="one">
      <div>
        <h4>content one</h4>
        <p>text</p>
      </div>
      {/* test */}
    </TabsContent>
    <TabsContent index="two">
      <div>
        <h4>content two</h4>
        <p>text</p>
      </div>
    </TabsContent>
  </Tabs>
);

export const SetDefaultTabs = () => (
  <div>
    <Tabs defaultTab="one" onClick={action('tabs selected')}>
      <TabsPanel>
        <Tab value="one" label="tab one"></Tab>
        <Tab value="two" label="tab two"></Tab>
      </TabsPanel>
      <TabsContent index="one">
        <div>
          <h4>content one</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="two">
        <div>
          <h4>content two</h4>
          <p>text</p>
        </div>
      </TabsContent>
    </Tabs>
    <hr />
    <Tabs defaultTab="two" onClick={action('tabs selected')}>
      <TabsPanel>
        <Tab value="one" label="tab one"></Tab>
        <Tab value="two" label="tab two"></Tab>
      </TabsPanel>
      <TabsContent index="one">
        <div>
          <h4>content one</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="two">
        <div>
          <h4>content two</h4>
          <p>text</p>
        </div>
      </TabsContent>
    </Tabs>
  </div>
);

export const DiffTypeTabs = () => (
  <div>
    <Tabs defaultTab="one" onClick={action('tabs selected')}>
      <TabsPanel>
        <Tab value="one" label="tab one"></Tab>
        <Tab value="two" label="tab two"></Tab>
      </TabsPanel>
      <TabsContent index="one">
        <div>
          <h4>content one</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="two">
        <div>
          <h4>content two</h4>
          <p>text</p>
        </div>
      </TabsContent>
    </Tabs>
    <hr />
    <Tabs defaultTab="one" onClick={action('tabs selected')}>
      <TabsPanel type="primary">
        <Tab value="one" label="tab one"></Tab>
        <Tab value="two" label="tab two"></Tab>
      </TabsPanel>
      <TabsContent index="one">
        <div>
          <h4>content one</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="two">
        <div>
          <h4>content two</h4>
          <p>text</p>
        </div>
      </TabsContent>
    </Tabs>
  </div>
);
export const DisabledTabs = () => (
  <div>
    <Tabs defaultTab="one" onClick={action('tabs selected')}>
      <TabsPanel>
        <Tab value="one" label="tab one"></Tab>
        <Tab value="two" label="tab two" disabled></Tab>
      </TabsPanel>
      <TabsContent index="one">
        <div>
          <h4>content one</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="two">
        <div>
          <h4>content two</h4>
          <p>text</p>
        </div>
      </TabsContent>
    </Tabs>
  </div>
);
export const WrappedTabs = () => (
  <div>
    <Tabs defaultTab="one" onClick={action('tabs selected')}>
      <TabsPanel>
        <Tab
          value="one"
          label="New Arrivals in the Longest Text of Nonfiction"
          wrapped
        ></Tab>
        <Tab value="two" label="tab two"></Tab>
      </TabsPanel>
      <TabsContent index="one">
        <div>
          <h4>content one</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="two">
        <div>
          <h4>content two</h4>
          <p>text</p>
        </div>
      </TabsContent>
    </Tabs>
  </div>
);
export const centeredTabs = () => (
  <div>
    <Tabs defaultTab="one" onClick={action('tabs selected')}>
      <TabsPanel centered>
        <Tab value="one" label="tab one"></Tab>
        <Tab value="two" label="tab two"></Tab>
      </TabsPanel>
      <TabsContent index="one">
        <div>
          <h4>content one</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="two">
        <div>
          <h4>content two</h4>
          <p>text</p>
        </div>
      </TabsContent>
    </Tabs>
  </div>
);
export const scrollableTabs = () => (
  <div>
    <Tabs defaultTab="one" onClick={action('tabs selected')}>
      <TabsPanel scrollable>
        <Tab value="one" label="tab one"></Tab>
        <Tab value="two" label="tab two"></Tab>
        <Tab value="three" label="tab three"></Tab>
        <Tab value="four" label="tab four"></Tab>
        <Tab value="five" label="tab five"></Tab>
        <Tab value="six" label="tab six"></Tab>
      </TabsPanel>
      <TabsContent index="one">
        <div>
          <h4>content one</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="two">
        <div>
          <h4>content two</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="three">
        <div>
          <h4>content three</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="four">
        <div>
          <h4>content four</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="five">
        <div>
          <h4>content five</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="six">
        <div>
          <h4>content six</h4>
          <p>text</p>
        </div>
      </TabsContent>
    </Tabs>
  </div>
);
export const verticalTabs = () => (
  <div>
    <Tabs vertical defaultTab="one" onClick={action('tabs selected')}>
      <TabsPanel scrollable>
        <Tab value="one" label="tab one"></Tab>
        <Tab value="two" label="tab two"></Tab>
        <Tab value="three" label="tab three"></Tab>
        <Tab value="four" label="tab four"></Tab>
        <Tab value="five" label="tab five"></Tab>
        <Tab value="six" label="tab six"></Tab>
      </TabsPanel>
      <TabsContent index="one">
        <div>
          <h4>content one</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="two">
        <div>
          <h4>content two</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="three">
        <div>
          <h4>content three</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="four">
        <div>
          <h4>content four</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="five">
        <div>
          <h4>content five</h4>
          <p>text</p>
        </div>
      </TabsContent>
      <TabsContent index="six">
        <div>
          <h4>content six</h4>
          <p>text</p>
        </div>
      </TabsContent>
    </Tabs>
  </div>
);
