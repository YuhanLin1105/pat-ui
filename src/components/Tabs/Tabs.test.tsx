import React from 'react';

import { render, fireEvent } from '@testing-library/react';
import Tabs from './Tabs';
import TabsPanel from './TabsPanel';
import Tab from './Tab';
import TabsContent from './TabsContent';

describe('Tabs', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<Tabs></Tabs>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render a default tabs container with tabsPabel and tabsContent based on user input', () => {
    const tabProps = {
      onClick: jest.fn(),
    };
    const wrapper = render(
      <Tabs>
        <TabsPanel>
          <Tab value="one" label="tab one"></Tab>
          <Tab value="two" label="tab two" {...tabProps}></Tab>
        </TabsPanel>
        <TabsContent index="one">
          <div>
            <h4>content one</h4>
            <p>text</p>
          </div>
        </TabsContent>
        <TabsContent index="two">
          <div>
            <h4>content two</h4>
            <p>text</p>
          </div>
        </TabsContent>
      </Tabs>
    ).container.firstChild;

    expect(wrapper).toHaveClass('tabs');
    expect(wrapper?.childNodes.length).toBe(3);
    expect(wrapper?.firstChild?.childNodes.length).toBe(2);
    expect(wrapper?.firstChild).toHaveClass('tabs__panel');
    expect(wrapper?.lastChild).toHaveClass('tabs__content');
    expect(tabProps.onClick).toBeCalledTimes(0);
    fireEvent.click(
      wrapper?.firstChild?.lastChild?.childNodes.item(0) as HTMLElement
    );
    expect(tabProps.onClick).toBeCalledTimes(1);
  });
  it('should render a correct tabs container based on different props and user input', () => {
    const tabsProps = {
      defaultTab: 'two',
      vertical: true,
      onClick: jest.fn(),
    };
     const tabProps = {
       onClick: jest.fn(),
     };
    const wrapper = render(
      <Tabs {...tabsProps}>
        <TabsPanel scrollable type="primary">
          <Tab value="one" label="tab one" disabled {...tabProps}></Tab>
          <Tab value="two" label="tab two"></Tab>
        </TabsPanel>
        <TabsContent index="one">
          <div>
            <h4>content one</h4>
            <p>text</p>
          </div>
        </TabsContent>
        <TabsContent index="two">
          <div>
            <h4>content two</h4>
            <p>text</p>
          </div>
        </TabsContent>
      </Tabs>
    ).container.firstChild;

    expect(wrapper).toHaveClass('vertical');
    expect(wrapper?.firstChild).toHaveClass(
      'vertical scrollable-col panel-type-primary'
    );
    expect(wrapper?.firstChild?.lastChild?.firstChild).toHaveClass(
      'actived-vertical'
    );
    //window.HTMLElement.prototype.scrollIntoView = function () {};
    expect(tabsProps.onClick).toHaveBeenCalledTimes(0);
    fireEvent.click(
      wrapper?.firstChild?.firstChild?.childNodes.item(0) as HTMLElement
    );
    expect(tabsProps.onClick).toHaveBeenCalledTimes(0);
  });
});
