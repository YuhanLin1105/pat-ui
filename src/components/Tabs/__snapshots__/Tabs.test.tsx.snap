// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttton should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="tabs"
  >
     
    <div
      class="tabs__panel panel-type-default"
    />
  </div>
</DocumentFragment>
`;

exports[`Tabs Renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="tabs"
  />
</DocumentFragment>
`;

exports[`Tabs should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="tabs"
  />
</DocumentFragment>
`;

exports[`Tabs should match snapshot 2`] = `
<DocumentFragment>
  <div
    class="tabs"
  />
</DocumentFragment>
`;
